### ISSUE TITLE
Controller: `invalid config map object received in config watcher. Ignored processing` with CPU spike

### ISSUE THREAD

**Issue by juranir** [argoproj/argo-workflows#11657]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [x] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

We're migrating our Argo workflow from v3.0.2 hosted in an old Rancher cluster to the last version in an AWS EKS cluster.

Ramdomically the workflow controller increases de CPU usage and starts to show this error.
`level=error msg="invalid config map object received in config watcher. Ignored processing"`

If we remove the pod, the new one starts well, but after 1 or 2 days (random), the error reappears.

The last one started yesterday at the moment without workflow running nor scheduled tasks on the infra side.

We don't have an error in the k8s API log to indicate some EKS outage.

The Argo server component doesn't have errors.

We don't have this error in the old version.

**Env:**
Argo: v3.4.10
Installation method: Official Helm Chart (0.33.1)
K8s Cluster: AWS EKS 1.27

---

**Workflow-controller k8s config map:**
> Name:         argo-workflow-controller-configmap
> Namespace:    argo
> Labels:       app.kubernetes.io/component=workflow-controller
>               app.kubernetes.io/instance=argo
>               app.kubernetes.io/managed-by=Helm
>               app.kubernetes.io/name=argo-workflows-cm
>               app.kubernetes.io/part-of=argo-workflows
>               helm.sh/chart=argo-workflows-0.33.1
> Annotations:  meta.helm.sh/release-name: argo
>               meta.helm.sh/release-namespace: argo
> 
> Data
> 
> config:
> 
> metricsConfig:
>   enabled: true
>   path: /metrics
>   port: 9090
>   ignoreErrors: false
>   secure: false
> persistence:
>   archive: true
>   archiveTTL: 60d
>   postgresql:
>     database: argo
>     host: <REDACTED>
>     passwordSecret:
>       key: password
>       name: argo-postgres-config
>     port: 5432
>     ssl: false
>     sslMode: disabled
>     tableName: argo_workflows
>     userNameSecret:
>       key: username
>       name: argo-postgres-config
> workflowDefaults:
>   spec:
>     ttlStrategy:
>       secondsAfterFailure: 604800
> sso:
>   issuer: https://<REDACTED>
>   clientId:
>     name: argo-server-clientid
>     key: clientid
>   clientSecret:
>     name: argo-server-clientsecret
>     key: clientsecret
>   redirectUrl: https://<REDACTED>/oauth2/callback
>   rbac:
>     enabled: true
>   scopes: 
>     - email
>     - profile
>     - openid
>   issuerAlias: https://<REDACTED>
>   sessionExpiry: 240h
> nodeEvents:
>   enabled: false`

  
---
**Resource Usage**
![image](https://github.com/argoproj/argo-workflows/assets/6254271/1833da65-24c5-4789-bffb-d5c5bede2f3e)

### Version

v3.4.10

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
N/A
Since it starts randomly, we usually don't have any workflows running
```

### Logs from the workflow controller

```text
Before:



----

After (Lines like that appear several times - I only have it in the container output):

time="2023-08-23T09:41:01.370Z" level=error msg="invalid config map object received in config watcher. Ignored processing"
```

### Logs from in your workflow's wait container

```text
N/A
```

**Comment by agilgur5** [comment]
> We don't have this error in the old version.

What was your old version? 3.4.9?
If it is a regression, knowing between which two versions would be very helpful

**Comment by juranir** [comment]
> What was your old version? 3.4.9?

No, this old version is v3.0.2.

This same behavior also happened with the 2 last versions (we made some downgrades to try to evaluate).

We just installed the last version to validate if everything would be smooth, so all workflows were good, it's the only issue we had.

**Comment by agilgur5** [comment]
Gotcha. Been a lot of changes since 3.0.2. Though I took a look and [this specific code](https://github.com/argoproj/argo-workflows/blame/82ac98026994b8b7b1a0486c6f536103d818fa99/workflow/controller/controller.go#L402)hasn't been changed in years since it was written.

Only thing I can potentially guess at is maybe your ConfigMap has some old config from 3.0.2 that is no longer valid in 3.4.10? Not sure why it wouldn't error earlier though

**Comment by juranir** [comment]
Thanks a lot for your help :D

I just compared my values with [chart values](https://github.com/argoproj/argo-helm/blob/main/charts/argo-workflows/values.yaml) and I don't have any unknown parameters.

Do you have any idea how could I debug more this error? Like figuring out the kind of object this code is receiving or any kind of data

**Comment by agilgur5** [comment]
Could try to run a debugger and inspect the variables then. Or could try to modify the code to log it out, but trying to log may itself fail.

**Comment by juranir** [comment]
I updated the code to try to register the "invalid object" that is causing the error. What is happening is that the **[event](https://github.com/argoproj/argo-workflows/blame/82ac98026994b8b7b1a0486c6f536103d818fa99/workflow/controller/controller.go#L400)** variable is null. I don't have another error log.
I'm using AWS EKS, so I also checked its log and I have no errors.

So I have no idea about this :(

**Comment by agilgur5** [comment]
Hmmm interesting find!

> `if event.Object == nil {`

Didn't you say that `event` _itself_ was nil?

> Maybe this [issue](https://github.com/kubernetes/client-go/issues/334) helps explain this behavior (I also found other complaints similar to this for projects other than Argo), if so I think we would add a conditional check [before](https://github.com/argoproj/argo-workflows/blame/82ac98026994b8b7b1a0486c6f536103d818fa99/workflow/controller/controller.go#L400) casting the object:

So in that issue they [suggest](https://github.com/kubernetes/client-go/issues/334#issuecomment-370561966) using a Reflector or Informer instead to handle several kinds of edge cases including this one. The codebase does use Informers very frequently already, so maybe we should just convert this one. The `RetryWatcher` [source code](https://github.com/kubernetes/client-go/blob/4783d283596145dd621e01636a829ec43b5101f9/tools/watch/retrywatcher.go#L48) does mention using Informers instead as well

**Comment by toblich** [comment]
We're running 3.4.9 and have seen this happen twice in the last month, where the controller starts logging multiple times per second the line ```level=error msg="invalid config map object received in config watcher. Ignored processing"``` and starts consuming a lot more resources, and the problems do not go away until we delete the pod. The config map itself is properly configured, and after a few days of running ok, it suddenly begins to spam that error log.

For context, the watcher that logs that error is actually watching the wrong namespace when both `--namespaced` and `--managed-namespace=<other-namespace>` are used (see https://github.com/argoproj/argo-workflows/issues/11463). I have no idea whether that is related to the nil event or not, but it leads to the watcher actually watching a non-existent config map.

**Comment by agilgur5** [comment]
> @agilgur5 I'm testing the code below, so far everything is fine, but I'm not sure if it's the best approach:

A bit hard to read as there isn't syntax highlighting and the indentation got screwed up a little, but that roughly looks correct to me. It also looks similar to other parts of the codebase.

Only thing I can think of is that `AddFunc` and `DeleteFunc` might be necessary for if the ConfigMap gets entirely deleted or added during runtime. But I'm not sure if the old code even handled that and it is an edge of an edge case, so not super critical.

Admittedly, I am not the most familiar with the internal workings here. If you want to submit a PR, we can certainly get others to give their feedback

> but I'm not sure if it's the best approach

Is there anything in particular that you were concerned about implementation-wise?

**Comment by juranir** [comment]
I waited a bit to see if the error would appear again. So far this code has solved the initial problem, but a new message has started to appear in the log (every 1min):
`listers.go:79] can not retrieve list of objects using index : Index with name namespace does not exist`


Even with this message, everything still works well, but I believe I made some mistakes that generated this error.

@agilgur5  or does anyone have an idea about that?


PS. Without my change this message will not appear

**Comment by agilgur5** [comment]
Not sure about the exact error message, but maybe the `duration` of `0` from your previous [code comment](https://github.com/argoproj/argo-workflows/issues/11657#issuecomment-1715378071) is too short?
Also there is a separate issue mentioned [above](https://github.com/argoproj/argo-workflows/issues/11657#issuecomment-1713990369) re: managed namespace that potentially impacts it. If you change it to use regular NS and that fixes it, then the error would be due to #11463

**Comment by roitalpaz** [comment]
We also just updated to 3.4.10 and see this problem as well.
any workarounds available?
Should we downgrade to 3.4.8?

**Comment by juranir** [comment]
I tried to increase the resync period and I also tried to define the namespace manually, but the message still appeared every minute.
I think this log message comes from [this code](https://github.com/kubernetes/client-go/blob/ee1a5aaf793a9ace9c433f5fb26a19058ed5f37c/tools/cache/listers.go#L55).

I'm not sure if it's a real error, but I think so because I hadn't this log before

I also tested with the latest version and the error still remains.

**Comment by juranir** [comment]
I'm starting to think I'm doing something wrong in the image construction process.

I tried cloning the repo and running the command `make workflow-controller-image` (without file changes), and the error message `listers.go:79] can not retrieve list of objects using index : Index with name namespace does not exist` keeps appearing, but if I use the official image it doesn't.

I tried using Windows and Mac M1(using --platform linux/amd64 parameter) as host.

PS. The build task ends without error.

**Comment by juranir** [comment]
There is some problem with the local environment (I don't know why). I downloaded the latest version code and made the same change. now the core is working fine without the weird message.

I'll wait some time and create a PR.

**Comment by agilgur5** [comment]
> We also just updated to 3.4.10 and see this problem as well.
> Should we downgrade to 3.4.8?

@roitalpaz this seems to have existed well before 3.4.10. as I mentioned [above](https://github.com/argoproj/argo-workflows/issues/11657#issuecomment-1690910568), this code path hasn't changed in years

I'd guess that occurrence was incidental as such, especially as there are no new features in 3.4.10 (if there was a configmap feature that might add up). Perhaps this issue just happens more frequently after Controller restarts

> any workarounds available?

as mentioned in the opening comment, restarting you Controller Pod will resolve this

**Comment by zzikkzzakk** [comment]
We are currently running into the same issue and it breaks our whole logging system.
Maybe it's in general a good idea to change the log level to Debug in places which could generate such high load to the logging system: https://github.com/argoproj/argo-workflows/blob/206c901c6ff122ad06a32820763e504cc5caf190/workflow/controller/controller.go#L414
?

**Comment by agilgur5** [comment]
> Maybe it's in general a good idea to change the log level to Debug

Log levels should be set based on their context. This is an error and so an error level is _correct_.
Errors are meant to be highly visible, you wouldn't want them to be silent or easily missed. Debug logs are mainly supposed to be helpful when debugging and not normally visible.

> in places which could generate such high load to the logging system

This is not necessarily something you can tell ahead of time. In this case, this is an edge case with `RetryWatcher`s that is apparently unhandled. The code did not expect that and so it logs an error. Because it logged this error a bug report was filed and it is now being fixed.

**Comment by andrii-shevchyk** [comment]
I've started observing same behaviour two days after I updated EKS 1.24 -> 1.28.
Argo workflow version: v3.3.5

**Comment by razielgn** [comment]
Running 3.5.1 on EKS 1.28.3, same issue.

**Comment by nickyfoster** [comment]
Seeing the same issue in EKS. Argo-controller is producing ~1.8 Mb/s of logs.

EKS: v1.28.3-eks-4f4795d
Argo controller: v3.4.4

**Comment by agilgur5** [comment]
> It seems this is only happening with kubernetes version 1.27 or later. Does anyone have evidence of it happening on earlier versions?

👀 I was wondering why we suddenly got more reports of  this.

> We are still using client-go from 0.24, which should not be compatible with 1.27, so perhaps that's the cause here.

The informer change also appears to fix it and we should still make that change, but if there's a simpler interim fix if that'd be great.

k8s is backward compatible with all clients (see also the [`client-go` compat docs](https://github.com/kubernetes/client-go#compatibility-client-go---kubernetes-clusters)), so I'm not so sure that upgrading `client-go` would fix it. It's possible they handled this `RetryWatcher` edge case better in newer versions if they've seen this more frequently in newer k8s versions, but one would think that would get backported as a fix per the `client-go` policy.

**Comment by tooptoop4** [comment]
maybe this k8s 1.27 change is related? https://github.com/kubernetes/kubernetes/pull/115620

**Comment by agilgur5** [comment]
> maybe this k8s 1.27 change is related? [kubernetes/kubernetes#115620](https://github.com/kubernetes/kubernetes/pull/115620)

that seems to be an Informer change, whereas the current code uses a `RetryWatcher` (it is being updated to an Informer in #11865)

**Comment by pcperera** [comment]
Is there a workaround for this issue ? Our logs are being flooded with the particular error.

**Comment by tooptoop4** [comment]
@pcperera try building argo binary with https://github.com/argoproj/argo-workflows/pull/12400 fix, since the watching code really belongs in argoCD not argoWorkflows

**Comment by agilgur5** [comment]
> since the watching code really belongs in argoCD not argoWorkflows

Argo CD can't restart Pods due to a ConfigMap change. That is the purpose of [Reloader](https://github.com/stakater/Reloader) though.
But the current approach is different from Reloader, which causes full restarts. Currently, the `RetryWatcher` actually causes a hot-update at run-time to the in-memory config with no restart required. Argo CD has a similar functionality for its own ConfigMaps as well.

> try building argo binary with #12400 fix

but yes that does contain a good interim workaround

**Comment by toyamagu-2021** [comment]
Should we recreate the retryWatcher instance if we got not ok from `ResultChan()`?
This might explain why we got an infinite log.

https://github.com/argoproj/gitops-engine/blob/aba38192fb16c35a6731717d81454a219565bcce/pkg/cache/cluster.go#L652-L660

**Comment by agilgur5** [comment]
That might make sense, since there is a [`continue`](https://github.com/argoproj/argo-workflows/blob/93914261cff4216561c89c1f5f6123e7ad0d5f61/workflow/controller/controller.go#L428) there currently. I would think that re-creating might result in the same problem though, but idk. I still don't know why this is happening more in k8s 1.27 specifically.

In either case though, we're looking to replace the `RetryWatcher` with a proper `Informer` in #11855 per discussion above as it handles certain edge cases better according to its docs.

**Comment by tooptoop4** [comment]
@juranir did u notice etcd size grow?

**Comment by mhehle** [comment]
Same issue on :
Kubernetes v1.27.7
Argo Worflows 3.5.2

argo workflow controller gets the error `invalid config map object received in config watcher. Ignored processing` sporadically and causes high CPU and IO utilization.

**Comment by tooptoop4** [comment]
maybe caused by https://github.com/kubernetes/kubernetes/issues/121688#issuecomment-1789860849

**Comment by agilgur5** [comment]
#12622 provides a workaround, not yet a full fix. I'll be taking over #11855 as OP [hasn't responded](https://github.com/argoproj/argo-workflows/pull/11855#issuecomment-1879765154) there in a bit, and that should (hopefully) be the full fix.

**Comment by agilgur5** [comment]
> maybe caused by [kubernetes/kubernetes#121688 (comment)](https://github.com/kubernetes/kubernetes/issues/121688#issuecomment-1789860849)

That's an interesting one. It does say that it doesn't change Informer behavior. But I'm not sure if that's the root cause since that wouldn't return this specific error -- it changes where it's retrieved from but not the contents (per se)

**Comment by stefansedich** [comment]
This one bit us ingesting 81TiB of logs into DataDog over a month resuling in some decent ingestion charges, we had an index exclusion filter however recently we had a larger spike in ingestion volume of this error across a number of instances.

Currently trying to filter this out at the logging agent so we stop sending these logs but a long term fix would be welcomed.

**Comment by sstarcher** [comment]
We are also seeing this after upgrading to K8S 1.27.  It's producing massive log spam and making our logging systems very angry.

**Comment by cdemarco-drw** [comment]
Same; can't tie to the upgrade but it's running on a K8s v1.27 cluster and spamming the heck outta Splunk.

**Comment by sstarcher** [comment]
We reverted back to an older version and we still have the same issue.  The only known change is upgrading from 1.26 to 1.27 K8S and this is producing 100gb + logs per day for every instance of argo workflows we have running.

**Comment by weretheone** [comment]
Hi @agilgur5 , do I understand well that this might be the permanent solution, for this problem and #12622 was more like a quick fix?
When do you expect this to be merged, released? It's worth to wait for this solution a bit more or should we go ahead with the other fix and not watch the configmap.

**Comment by juranir** [comment]
@agilgur5,  I'm so sorry for my delay.
I had some unforeseen circumstances personal/professional but now I'm back and, if possible, I would like to finish this PR.

**Comment by NatElkins** [comment]
Just want to chime in and say we are also experiencing this issue.  Thank you for your work @agilgur5 and @juranir!!  I'm not familiar with the internals of Argo but if there's anything we can do to test things out (using a pre-release, etc) please let me know.

**Comment by kgyovai** [comment]
I'm also experiencing this issue. I did a major K8s upgrade from v 1.24 to 1.29 on 03/25/24. No issues for almost __exactly__ 7 days. Then at what appears to be very close to the exact minute of 7 days after the upgrade, I started receiving CloudWatch Billing alerts on 04/01/24. After tracking those down, it was clear that the cause was a 100X spike in log volume (CloudWatch measures in bytes).

<img width="1426" alt="image" src="https://github.com/argoproj/argo-workflows/assets/10728023/926684b5-a348-4adb-815a-3249c6ebe117">

It's odd that this happened nearly exactly 7 days after the upgrade. Logs were stable for the first 2-3 hours after midnight on 04/01/24 - which indicates that the issue wasn't caused by a new month/day/week.

**Comment by agilgur5** [comment]
> but if there's anything we can do to test things out (using a pre-release, etc) please let me know.

@NatElkins if you and anyone else can test #11855, that would be great! I pushed up [`agilgur5/argo-workflow-controller:fix-configmap-informer`](https://hub.docker.com/layers/agilgur5/argo-workflow-controller/fix-configmap-informer/images/sha256-48c229a153ff18c9ab6fc94b4f62340e6155a6d68c25937c4ce733f73b444345?context=repo) for testing

**Comment by agilgur5** [comment]
Thanks for the quick response @dcerbino 

> However, I believe it's prudent to first update our Helm chart to version 0.41.1

It's not strictly necessary. The main parts in the Helm [changelog](https://artifacthub.io/packages/helm/argo/argo-workflows?modal=changelog) are the past few Argo patch updates.
That would make for less differences that could be confounding variables though. But if you're just verifying the CPU and logs, the other changes shouldn't have significant impact on either.

> configure the `controller.image.tag` to `fix-configmap-informer`

The image is uploaded to my personal DockerHub, so more than that: also `registry` to `""` and `repository` to `agilgur5/argo-workflow-controller`

**Comment by dcerbino** [comment]
@agilgur5 the controller is running but got some strange logs:
`E0409 15:23:24.840538       1 reflector.go:140] pkg/mod/k8s.io/client-go@v0.26.15/tools/cache/reflector.go:169: Failed to watch *v1.ConfigMap: failed to list *v1.ConfigMap: configmaps "argo-workflows-workflow-controller-configmap" is forbidden: User "system:serviceaccount:argo:argo-workflows-controller" cannot list resource "configmaps" in API group "" at the cluster scope`

I will attach the logs.

Is it expected? 
[argo_controller.log](https://github.com/argoproj/argo-workflows/files/14920708/argo_controller.log)

**Comment by agilgur5** [comment]
No, that would suggest your SA does not have the proper `ClusterRole`; missing `list` `configmaps`? If you're installing with the Helm Chart, it is included [by default with `configmaps` RBAC](https://github.com/argoproj/argo-helm/blob/af50be59ed31a371f811ad586bd72a762291b810/charts/argo-workflows/templates/controller/workflow-controller-cluster-roles.yaml#L37)

**Comment by dcerbino** [comment]
Hello @agilgur5, Thx for your help.
I've manually rollback to standard helm release. No role error in this case.
Then I've edited the deployment of argo controller to pull your image and I've again logs error.
So it seams the the role, SA and binding is ok but for some reason your fix fail to read list the configmap. 

I will attach SA role and rolebinding.
[rolebinding.yaml.txt](https://github.com/argoproj/argo-workflows/files/14932789/rolebinding.yaml.txt)
[sa.yaml.txt](https://github.com/argoproj/argo-workflows/files/14932790/sa.yaml.txt)
[role.yaml.txt](https://github.com/argoproj/argo-workflows/files/14932792/role.yaml.txt)

**Comment by agilgur5** [comment]
Hmm, that is a `Role`, not a `ClusterRole`. Are you using a managed namespace install? Could you provide some of your Helm `values` so I know the configuration?

With a `Role`, that error makes sense, your Controller indeed does not have _cluster-scope_ permissions.

But the code should be using your managed namespace if it exists; cluster-scope is the default when it _doesn't_ exist.

**Comment by dcerbino** [comment]
HI @agilgur5 we use argo as namespace for argo staff.
here is our argo-workflows.yaml.tpl that we use as values for the helm chart
`images:
  pullPolicy: IfNotPresent


singleNamespace: true

workflow:
  serviceAccount:
    create: true # Specifies whether a service account should be created
    name: "${service_account_name}"
    annotations:
      eks.amazonaws.com/role-arn: "${role_arn}"

  rbac:
    create: true

controller:
  containerRuntimeExecutor: emissary
  # -- The duration in seconds before the pods in the GC queue get deleted. A zero value indicates that the pods will be deleted immediately.
  # -- we need this to give new relic agent time to ingest logs
  podGCDeleteDelayDuration: "600s"
  nodeSelector:
    pool: monitor
  metricsConfig:
    enabled: true

  parallelism: 10

  podAnnotations:
    cluster-autoscaler.kubernetes.io/safe-to-evict: "true"
  image:
    tag: "fix-configmap-informer"
    repository: "agilgur5/argo-workflow-controller"
    registry: ""

  workflowDefaults:
    spec:
      serviceAccountName: "${service_account_name}"
      imagePullSecrets:
        - name: regcred
      tolerations:
        - key: "dedicated"
          operator: "Equal"
          value: "spot"
          effect: "NoSchedule"

  serviceMonitor:
    enabled: true
    additionalLabels: {}

  serviceAccount:
    create: true
    name: "${service_account_name}-controller"
  instanceID:
    enabled: false

  logging:
    level: info
    globallevel: "0"
    format: "json"

  resources:
    requests:
      cpu: 100m
      memory: 200Mi

  replicas: 1
  pdb:
    enabled: true
    minAvailable: 1
    maxUnavailable: 1

  clusterWorkflowTemplates:
    # Create a ClusterRole and CRB for the controller to access ClusterWorkflowTemplates.
    enabled: false

server:
  name: server
  nodeSelector:
    pool: monitor
  logging:
    format: "json"
  podAnnotations:
    cluster-autoscaler.kubernetes.io/safe-to-evict: "true"

  serviceAccount:
    create: true
    name: "${service_account_name}-server"

  resources:
    requests:
      cpu: 100m
      memory: 200Mi

  replicas: 2

  pdb:
    enabled: true
    minAvailable: 1
    maxUnavailable: 1


  # Run the argo server in "secure" mode. Configure this value instead of
  # "--secure" in extraArgs. See the following documentation for more details
  # on secure mode:
  # https://argoproj.github.io/argo-workflows/tls/
  # secure: true

  extraArgs:
    - "--auth-mode=server"
`
I hope this can help.

**Comment by agilgur5** [comment]
Thanks @dcerbino.
I found the bug, there's apparently a not-well-known config (so much so that I forgot about it myself after I added missing docs for it in #11297) in the `workflow-controller-configmap` for [`config.namespace`](https://github.com/argoproj/argo-workflows/blob/fc30b5a2b790c99b8c5addf67d14dbbefea7418d/docs/workflow-controller-configmap.yaml#L16). That's [not available](https://github.com/argoproj/argo-helm/blob/af50be59ed31a371f811ad586bd72a762291b810/charts/argo-workflows/templates/controller/workflow-controller-config-map.yaml#L10) in the Helm Charts (probably because it hadn't been documented), so it's coming up as an empty string on your install, and am empty `namespace` selector means all namespaces/cluster-scope.
This `config.namespace` can be different from Argo's own `namespace` and the managed namespace, whereas I mistakenly thought it was the same as Argo's own `namespace` (I thought it was the configmap's namespace). 

I'll push a patch for that shortly, thanks for testing and providing config info!

**Comment by agilgur5** [comment]
> I'll push a patch for that shortly, thanks for testing and providing config info!

pushed up [`agilgur5/argo-workflow-controller:fix-configmap-informer-v2`](https://hub.docker.com/layers/agilgur5/argo-workflow-controller/fix-configmap-informer-v2/images/sha256-f553ce51e22a126a99705563d3317527e2da70478d19f976db6093d299f296b5?context=repo) for testing

I also overwrote `agilgur5/argo-workflow-controller:fix-configmap-informer` to the new version

**Comment by dcerbino** [comment]
Thanks @agilgur5.
I've just pulled new image and no error in the log now.
Again thanks for your help.
I will let you now about the original bug.

**Comment by tooptoop4** [comment]
maybe this is relevant? https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.30.md#bug-or-regression-2 mentions:
 "Kube-apiserver: fixes a 1.27+ regression in watch stability by serving watch requests without a resourceVersion from the watch cache by default, as in <1.27 (disabling the change in [#115096](https://github.com/kubernetes/kubernetes/pull/115096) by default). This mitigates the impact of an etcd watch bug (https://github.com/etcd-io/etcd/pull/17555). If the 1.27 change in #115096 to serve these requests from underlying storage is still desired despite the impact on watch stability, it can be re-enabled with a WatchFromStorageWithoutResourceVersion feature gate. (https://github.com/kubernetes/kubernetes/pull/123935, [@serathius](https://github.com/serathius)) [SIG API Machinery and Etcd]"

**Comment by agilgur5** [comment]
It might be, you mentioned the original change [previously](https://github.com/argoproj/argo-workflows/issues/11657#issuecomment-1935075776) as well. If `etcd` were serving some corrupted data back or something, then it would resolve that by reverting to the previous behavior.

**Comment by dcerbino** [comment]
Hello @agilgur5
It seams the the fix is working well.
We got no error from the new controller in our stage env.

**Comment by agilgur5** [comment]
Thanks for testing @dcerbino!
How's your memory and CPU usage compared to before? The error caused a CPU spike before which should now be gone. The fix should also require a bit more memory (although I optimized it quite heavily).
And to confirm, you're on k8s 1.27+ right?

**Comment by dcerbino** [comment]
HI @agilgur5 
I can confirm that we are running it in kubernetes 1.29.
No significant cpu change.
Memory change was about 3-4 Mbytes before 111 Mbyte after 115 Mbyte.
Even in this case not a significant change.

**Comment by agilgur5** [comment]
Thanks for providing those metrics! Awesome! That's actually better memory than I anticipated, although it is proportional to number of ConfigMaps so a bit dependent on cluster (the biggest optimization I made was to make the cache only store `name` and `namespace` of ConfigMaps where possible, which should only be a few bytes each)

**Comment by dcerbino** [comment]
Hello @agilgur5 
do you have any idea on when the fix will be available on the official helm chart?

**Comment by NatElkins** [comment]
Hi @agilgur5, this fix seems to be working perfectly, thank you for your efforts!

Do you know what additional review/steps are required to get this into main?

Thanks again!

**Comment by agilgur5** [comment]
> do you have any idea on when the fix will be available on the official helm chart?

The [official Helm Charts](https://github.com/argoproj/argo-helm) have separate maintainers, so I can't tell you for them. They generally update pretty quickly and I think have `renovate` doing automatic dep updates now: https://github.com/argoproj/argo-helm/pull/2651

They also only release patches for the latest minor (as it requires having separate release branches), unless that's changed recently. So only for 3.5.x right now.

Although you don't need the Helm Chart to update to update your image version, you can specify it in your `values.yaml`. #11855 also makes no CRD changes.

I'm currently the release manager for 3.5.x, so as soon as #11855 is merged, it'll get backported to the `release-3.5` branch, which releases every ~2-4 weeks

> Do you know what additional review/steps are required to get this into main?

The ones listed [in the PR](https://github.com/argoproj/argo-workflows/pull/11855#issuecomment-2054832642).

See also https://justinmayer.com/posts/any-updates/, https://sindresorhus.com/blog/issue-bumping

**Comment by tooptoop4** [comment]
@dcerbino would u be able to try the vanilla argo release on at least k8s 1.27.13 or 1.28.9 or 1.29.4 or 1.30.x? the k8s regression fix was backported from these minor versions

**Comment by dcerbino** [comment]
Hello @tooptoop4 
Sorry for late answer.
We have just installed the version 3.5.8 on our stage env. 
I will let you know about the bug fix as soon as possible.

**Comment by dcerbino** [comment]
@tooptoop4 and @agilgur5 
We have just bumped up argo workflow in prod. No error in stage until now. So this bug seems to be fixed with helm chart version 0.41.11 and argo workflow v3.5.8 in K8s Rev: v1.29.4.

**Comment by agilgur5** [comment]
> K8s Rev: v1.29.4.

Thanks for testing!

So that would confirm that this issue is due to an upstream bug that @tooptoop4 [mentioned](https://github.com/argoproj/argo-workflows/issues/11657#issuecomment-1915393192) [above](https://github.com/argoproj/argo-workflows/issues/11657#issuecomment-2050434984)

We should still get the PR in, it's a good change nonetheless, but this is due to a k8s bug [in 1.27](https://github.com/argoproj/argo-workflows/issues/11657#issuecomment-1835920034) that is fixed upstream [in 1.27.13+, 1.28.9+, 1.29.4+, and 1.30+](https://github.com/argoproj/argo-workflows/issues/11657#issuecomment-2215367233)

**Comment by weretheone** [comment]
Hello @agilgur5 ! The above would mean that this will be merged eventually?

**Comment by tooptoop4** [comment]
i tink dis can be closed as long as u keep your k8s patch version above 1.29.4.... 1.28 and earlier is already past EOL

**Comment by rossigee** [comment]
FTR, this is still happening for me with latest versions of all components.

After a few hours, the controller starts rapidly logging the error, consuming excessive CPU and network resources until the pod is killed.

Kubernetes: 1.32.4
Argo-Workflows: argoproj/workflow-controller:v3.6.7

**Comment by Reversaidx** [comment]
The same issue happened with me as well

**Comment by tooptoop4** [comment]
is this still an issue?

**Comment by ianwremmel** [comment]
I needed to add the following a month ago:

```
          # Workaround for high CPU bug: https://github.com/argoproj/argo-workflows/issues/11657
          extraEnv:
            - name: WATCH_CONTROLLER_SEMAPHORE_CONFIGMAPS
              value: "false"
```

but to be fair, it's in a homelab that claude doesn most of the maintenance. I'm still a k8s rookie :)

**Comment by weretheone** [comment]
> is this still an issue?

We are using modern Argo Workflows and on 3.7.8 we did not notice any issue.

**Comment by davidB** [comment]
> is this still an issue?
  
Yes, still reproducible as of April 2026.

  Environment:
  - Kubernetes: v1.34.6+k3s1 (K3s)
  - Argo Workflows app version: v4.0.4
  - Helm chart: argo-helm/argo-workflows v1.0.10 (also observed on v1.0.7)
  - Deployed via ArgoCD (GitOps)

  Symptom:
  The workflow-controller pod enters a tight CPU loop spamming component=configmap_watcher errors at > 3 lines/second:

  time=2026-04-19T10:04:26.220Z level=ERROR msg="invalid config map object received in semaphore config watcher. Ignored processing" component=configmap_watcher

  Context:
  - No workflows using semaphore synchronization are defined
  - Only 2 ConfigMaps exist in the namespace: the controller configmap and kube-root-ca.crt
  - Controller SA has correct watch configmaps RBAC permissions
  - No network policies blocking controller → API server
  - Pod never recovers — requires manual restart, which also immediately re-enters the loop

  Note on K8s version: The mentioned K8s 1.27+ upstream fix does not apply here — K3s v1.34.6 is well past that threshold, yet the bug persists. This appears to be entirely within the Argo Workflows controller code, not a K8s-side issue.

**Comment by hellekin08** [comment]
We are facing the same issue:
- Argo Workflows app version: v4.0.5
- Helm chart: 1.0.13
- Kubernetes:  AKS 1.33.5

Also no workflows using semaphore synchronization are defined.
